### PR TITLE
feat: [1/2] Top-k and Top-p support for dtensor worker with vLLM V0 when `TP==1`

### LINKED ISSUE #69: Top-p/Top-k Sampling Params handling in VLLM v1

### PR DESCRIPTION
**Pull request by zhandaz**
# What does this PR do ?

tldr: Support top-k and top-p for dtensor worker with vLLM v0. This pr only supports `tp==1`, the pr for `tp>1` is in #774.

1. By default, Nemo-RL is using vLLM v1. According to https://docs.vllm.ai/en/v0.8.1/getting_started/v1_user_guide.html#semantic-changes-and-deprecated-features, vLLM v1 returns the model’s raw output (i.e. before applying any logits post-processing such as temperature scaling or penalty adjustments). Therefore, we don't need to support temperature, top-k, or top-p.
2. **If we want to support Top-k and Top-p in vLLM v0, then it means the training logic of using vLLM v0 and v1 would be different. I.e., in v0, the loss is calculated from the processed logits while in v1 the loss is calculated from the raw logits.**
3. This pr assumes we want to make it possible. But supporting Top-k and Top-p with `tp_size > 1` would need changes on the distributed functions. Because Top-k/top-p filtering requires seeing the full vocabulary to determine which tokens to keep. From what I can tell, we need to change `DistributedLogprob` or `_compute_distributed_log_softmax` to support p and k. We have another PR for this case so that people can easily revert if the changes on distributed functions are considered not elegant.

For the detailed implementation, we add our implementation based on vLLM's: https://github.com/vllm-project/vllm/blob/34a20c49b3f81f64133428b3a0d62309db1256f9/vllm/v1/sample/ops/topk_topp_sampler.py.

# Issues

Related Issue: https://github.com/NVIDIA-NeMo/RL/issues/69

### CODE REVIEW

**Comment by SahilJain314** [comment]
thanks for adding this! With the upcoming vllm 0.10.0, we should be able to get v1 to output post-processed logprobs as well.

**Comment by Dazz993** [comment]
@SahilJain314 Thanks for pointing this out. In that case, I will wait for the pr for upgrading to `vllm=0.10.0` to be merged, and then I will add the option to choose the log processing methods.

File: nemo_rl/models/policy/utils.py [hunk 1]
```diff
@@ -14,14 +14,108 @@
...
+
+    # Apply top-p
+    probs_sort = logits_sort.softmax(dim=-1)
+    probs_sum = torch.cumsum(probs_sort, dim=-1)
+    top_p_mask = probs_sum <= 1 - top_p
+    # at least one - but for p=0.0, we want exactly one
+    if top_p == 0.0:
+        # Keep only the highest probability token
+        top_p_mask = torch.ones_like(top_p_mask, dtype=torch.bool)
+        top_p_mask[..., 0] = False  # Keep only the first (highest prob) token
```
[the remark this answers is not in the document]
Comment (zhandaz) [reply nemo_rl/models/policy/utils.py]: No. We use `descending=False`.

File: nemo_rl/models/policy/utils.py [hunk 2]
```diff
@@ -14,14 +14,108 @@
...
+        Filtered logits with sampling parameters applied
+    """
+    if top_p is None or top_p == 1.0:
+        if top_k is None or top_k == -1:
+            return logits
+        # Avoid sorting vocab for top-k only case
+        return apply_top_k_only(logits, top_k)
+
+    # Apply top-p (requires sorting)
+    logits_sort, logits_idx = logits.sort(dim=-1, descending=False)
```
[the remark this answers is not in the document]
Comment (zhandaz) [reply nemo_rl/models/policy/utils.py:51]: No. We keep `descending=False` to align with `vllm`.

File: nemo_rl/models/policy/utils.py [hunk 3]
```diff
@@ -14,14 +14,108 @@
...
+    top_p_mask = probs_sum <= 1 - top_p
+    # at least one - but for p=0.0, we want exactly one
+    if top_p == 0.0:
+        # Keep only the highest probability token
+        top_p_mask = torch.ones_like(top_p_mask, dtype=torch.bool)
+        top_p_mask[..., 0] = False  # Keep only the first (highest prob) token
+    else:
+        # at least one
+        top_p_mask[..., -1] = False
+    logits_sort.masked_fill_(top_p_mask, -float("inf"))
```
[the remark this answers is not in the document]
Comment (zhandaz) [reply nemo_rl/models/policy/utils.py]: No. We use `descending=False` to align with `vllm`.

**Comment by Dazz993** [comment]
Update: I found this PR to get the final logprobs from vllm: https://github.com/vllm-project/vllm/pull/22387

**Comment by 22quinn** [comment]
> Update: I found this PR to get the final logprobs from vllm: [vllm-project/vllm#22387](https://github.com/vllm-project/vllm/pull/22387)

Thank you for the interest! That PR will apply not only top-k & top-p, but also temperature & min-p. Does that work for your use case?

**Comment by Dazz993** [comment]
> > Update: I found this PR to get the final logprobs from vllm: [vllm-project/vllm#22387](https://github.com/vllm-project/vllm/pull/22387)
> 
> Thank you for the interest! That PR will apply not only top-k & top-p, but also temperature & min-p. Does that work for your use case?

Yep! We want the final logprobs just before the sampling. Thank you for your PR!

**Comment by wangshangsam** [comment]
@zhandaz Just to confirm - the state of this PR right now is that it's waiting for the vLLM 0.11 release?

**Comment by yuki-97** [comment]
closes since https://github.com/NVIDIA-NeMo/RL/pull/2053 merged.
